iris-mpc-gpu/src/server/actor.rs

```diff
@@ -1902,11 +1927,117 @@ impl ServerActor {
             return;
         }
 
-        // which database are we querying against
-        let (code_db_slices, mask_db_slices) = match eye_db {
-            Eye::Left => (&self.left_code_db_slices, &self.left_mask_db_slices),
-            Eye::Right => (&self.right_code_db_slices, &self.right_mask_db_slices),
-        };
+        if eye_db != self.full_scan_side && self.on_demand_loader.is_some() {
+            // load the database on demand
+            let on_demand_loader = self.on_demand_loader.as_mut().unwrap();
+
+            // map in-memory GPU indices to DB indices
+            let num_devices = db_subset_idx.len();
+            let db_indices = db_subset_idx
+                .iter()
+                .enumerate()
+                .flat_map(|(i, x)| x.iter().map(move |&y| y as usize * num_devices + i))
+                .collect::<Vec<_>>();
+            tracing::info!("{} DB indices: {:?},", self.party_id, db_indices);
+
+            // fetch them from the DB
+            let now = Instant::now();
+            // since we are in a tokio::spawn_blocking context, we can use the tokio runtime handle
+            let iris_codes = tokio::runtime::Handle::current().block_on(async {
+                on_demand_loader
+                    .load_records(eye_db, &db_indices)
+                    .await
+                    .expect("Can load records from on-demand loader")
```

Comment:
note: theoretically we could start streaming in those as we find them in the previous search. But this would also require how fetch those partial matches. Currently we fetch them after having processed all chunks, we could do that instead for every chunk and stream those already. Same for the `OR` indices, they are known at the start already.